client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,559 @@
+					*state = PeerState::NotConnected;
+					*peer_id
+				})
+			})
+			.collect::<Vec<_>>()
+			.iter()
+			.for_each(|peer_id| self.drop_connection(*peer_id));
+	}
+
+	fn on_add_peer(&mut self, peer_id: PeerId) {
+		if self.reserved_nodes.contains_key(&peer_id) {
+			return
+		}
+
+		if self.nodes.insert(peer_id, PeerState::NotConnected).is_none() {
+			self.alloc_slots();
+		}
+	}
+
+	fn on_remove_peer(&mut self, peer_id: PeerId) {
+		// Don't do anything if the node is reserved.
+		if self.reserved_nodes.contains_key(&peer_id) {
+			return
+		}
+
+		match self.nodes.remove(&peer_id) {
+			Some(state) => match state {
+				PeerState::Connected(d) => {
+					match d {
+						Direction::Inbound => self.num_in -= 1,
+						Direction::Outbound => self.num_out -= 1,
+					}
+					self.drop_connection(peer_id);
+				},
+				PeerState::NotConnected => {},
+			},
+			None => {
+				trace!(
+					target: "peerset",
+					"Trying to remove unknown peer {} from {:?}",
+					peer_id, self.set_id,
+				);
+			},
+		}
+	}
+
+	fn on_incoming_connection(&mut self, peer_id: PeerId, incoming_index: IncomingIndex) {
+		if self.reserved_only && !self.reserved_nodes.contains_key(&peer_id) {
+			self.reject_connection(incoming_index);
+			return
+		}
+
+		// Check if the node is reserved first.
+		if let Some(state) = self.reserved_nodes.get_mut(&peer_id) {
+			match state {
+				// If we're already connected, don't answer, as the docs mention.
+				PeerState::Connected(_) => {},
+				PeerState::NotConnected => {
+					// It's questionable whether we should check a reputation of reserved node.
+					// FIXME: unable to call `self.is_banned()` because of borrowed `self`.
+					if self.peerset_handle.is_banned(peer_id) {
+						self.reject_connection(incoming_index);
+					} else {
+						*state = PeerState::Connected(Direction::Inbound);
+						self.accept_connection(incoming_index);
+					}
+				},
+			}
+			return
+		}
+
+		// Adding incoming peer to our set of peers even before we decide whether to accept
+		// it is questionable, but this is how it was implemented in the original `Peerset`.
+		let state = self.nodes.entry(peer_id).or_insert(PeerState::NotConnected);
+		match state {
+			// If we're already connected, don't answer, as the docs mention.
+			PeerState::Connected(_) => return,
+			PeerState::NotConnected => {
+				// FIXME: unable to call `self.is_banned()` because of borrowed `self`.
+				if self.peerset_handle.is_banned(peer_id) {
+					self.reject_connection(incoming_index);
+					return
+				}
+
+				if self.num_in >= self.max_in {
+					self.reject_connection(incoming_index);
+					return
+				}
+
+				self.num_in += 1;
+				*state = PeerState::Connected(Direction::Inbound);
+				self.accept_connection(incoming_index);
+			},
+		}
+	}
+
+	/// Returns Err(PeerId) if peer wasn't connected.
+	fn on_peer_dropped(&mut self, peer_id: PeerId, reason: DropReason) -> Result<(), PeerId> {
+		let mut is_reserved = true;
```

Comment:
I think there could be two private methods that this method would call: one for reserved peers and one for normal peers. Having a mutable boolean flag that is used to decide which code path is taken seems unnecessary. Your call though

Reply:
Done